cmd/skopeo/sync.go

```diff
@@ -368,61 +378,158 @@ func imagesToCopyFromRegistry(registryName string, cfg registrySyncConfig, sourc
+}
+
+// filterFunc is a function used to limit the initial set of image references
+// using tags, patterns, semver, etc.
+type filterFunc func(*logrus.Entry, []types.ImageReference) []types.ImageReference
+
+// filterCollection is a map of image names to filter functions.
+type filterCollection map[string]filterFunc
+
+// filterSourceReferences lists tags for images specified in the collection and
+// filters them using assigned filter functions.
+// It returns a list of repoDescriptors.
+func filterSourceReferences(sys *types.SystemContext, registryName string, collection filterCollection) []repoDescriptor {
+	var repoDescList []repoDescriptor
+	for imageName, filter := range collection {
+		logger := logrus.WithFields(logrus.Fields{
 			"repo":     imageName,
 			"registry": registryName,
 		})
 		repoRef, err := parseRepositoryReference(fmt.Sprintf("%s/%s", registryName, imageName))
 		if err != nil {
-			repoLogger.Error("Error parsing repository name, skipping")
+			logger.Error("Error parsing repository name, skipping")
 			logrus.Error(err)
 			continue
 		}
 
-		repoLogger.Info("Processing repo")
+		logger.Info("Processing repo")
 
 		var sourceReferences []types.ImageReference
 
-		tagReg, err := regexp.Compile(tagRegex)
+		logger.Info("Querying registry for image tags")
+		sourceReferences, err = imagesToCopyFromRepo(sys, repoRef)
 		if err != nil {
-			repoLogger.WithFields(logrus.Fields{
-				"regex": tagRegex,
-			}).Error("Error parsing regex, skipping")
+			logger.Error("Error processing repo, skipping")
 			logrus.Error(err)
 			continue
 		}
 
-		repoLogger.Info("Querying registry for image tags")
-		allSourceReferences, err := imagesToCopyFromRepo(serverCtx, repoRef)
-		if err != nil {
-			repoLogger.Error("Error processing repo, skipping")
-			logrus.Error(err)
+		sourceReferences = filter(logger, sourceReferences)
+		if len(sourceReferences) == 0 {
+			logger.Warnf("No refs to sync found")
 			continue
 		}
 
-		repoLogger.Infof("Start filtering using the regular expression: %v", tagRegex)
-		for _, sReference := range allSourceReferences {
-			tagged, isTagged := sReference.DockerReference().(reference.Tagged)
-			if !isTagged {
-				repoLogger.Errorf("Internal error, reference %s does not have a tag, skipping", sReference.DockerReference())
-				continue
+		repoDescList = append(repoDescList, repoDescriptor{
+			ImageRefs: sourceReferences,
+			Context:   sys,
+		})
+	}
+	return repoDescList
+}
+
+// tagRegexFilterCollection converts a map of (image name, tag regex) pairs
+// into a filterCollection, which is a map of (image name, filter function)
+// pairs.
+func tagRegexFilterCollection(collection map[string]string) (filterCollection, error) {
+	filters := filterCollection{}
+
+	for imageName, tagRegex := range collection {
+		pattern, err := regexp.Compile(tagRegex)
+		if err != nil {
+			return nil, err
+		}
+
+		f := func(logger *logrus.Entry, sourceReferences []types.ImageReference) []types.ImageReference {
```

Comment:
This would be simpler if the filter only worked on a single reference, maybe `func(*logrus.Entry, types.ImageReference) bool`. The filtering loop could then exist only in `filterSourceReferences`

(It’s fine to lose the detailed “Start filtering using the regular expression: $value” to get this simplification; or alternatively `filterCollection` could be a `map[string]struct{filter func …, message string}`, but I don’t think that’s really necessary.

Really the “filtering” phase should take negligible time and I don’t think it needs a separate _progress_ log message at all. If the purpose of that log message were just to log the configuration that was used, that kind of logging could happen during the filter creation — but I wouldn’t bother with that without a specific indication of user need either.)

Reply:
That simplifies matters, indeed. I also think it's worth departing from logging all the details if it makes the code more difficult to navigate.